packages/transaction-preview/src/index.ts

```diff
@@ -0,0 +1,259 @@
+  TransactionPreviewAPI,
+  TransactionObject
+} from './types.js'
+import type { EIP1193Provider } from '@web3-onboard/common'
+
+import {
+  validateSetContainerEl,
+  validateSimTransactions,
+  validateTPInit
+} from './validation'
+import TransactionPreview from './views/Index.svelte'
+import initI18N from './i18n/index.js'
+import simulateTransactions from './simulateTransactions.js'
+
+export * from './types.js'
+
+const approved$ = new Subject<boolean>()
+let options: TransactionPreviewInitOptions
+let app: TransactionPreview
+
+export const setContainerElement = (containerElement: string): void => {
+  if (containerElement) {
+    const error = validateSetContainerEl(containerElement)
+
+    if (error) {
+      throw error
+    }
+  }
+  options.containerElement = containerElement
+}
+
+const destroyApp = () => {
+  app.$destroy()
+}
+
+const simTransactions = (
+  txs: [TransactionObject]
+): Promise<SimPlatformResponse> => {
+  if (txs) {
+    const error = validateSimTransactions(txs)
+
+    if (error) {
+      throw error
+    }
+  }
+  return simulateTransactions(options, txs)
+}
+
+const handleRequireApproval = async (
+  app: TransactionPreview,
+  fullProviderRequest: EIP1193Provider['request'],
+  req: {
+    method: string
+    params?: Array<unknown>
+  }
+) => {
+  const approved = await firstValueFrom(approved$)
+  app.$destroy()
+  if (!approved) {
+    throw new ProviderRpcError({
+      code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
+      message: 'User rejected the transaction'
+    })
+  }
+  fullProviderRequest(req)
+}
+
+export const patchProvider = (
+  walletProvider: PatchedEIP1193Provider
+): PatchedEIP1193Provider => {
+  if (!walletProvider) {
+    throw new Error(
+      `An EIP 1193 wallet provider is required to preform patching and 
+      watch for transactions e.g. an injected wallet using window.ethereum`
+    )
+  }
+  if (walletProvider.simPatched) return walletProvider as PatchedEIP1193Provider
+
+  const fullProviderRequest = walletProvider.request
+  const patchedProvider = walletProvider as PatchedEIP1193Provider
+  const request: EIP1193Provider['request'] = async (req: {
+    method: string
+    params?: Array<unknown>
+  }): Promise<any> => {
+    if (
+      req.method === 'eth_sendTransaction' &&
+      req.params &&
+      req.params.length
+    ) {
+      const transactionParams =
+        req.params as EthSignTransactionRequest['params']
+      try {
+        const preview = await simulateTransactions(options, transactionParams)
+        if (preview.status !== 'simulated') {
+          // If transaction simulation was unsuccessful do not create DOM el
+          return
+        }
+        if (app) app.$destroy()
+        app = mountTransactionPreview(preview)
+        options.requireTransactionApproval
+          ? handleRequireApproval(app, fullProviderRequest, req)
+          : fullProviderRequest(req).then((hash: unknown) => {
+              hash && app.$destroy()
```

Comment:
Do we want to destroy the Sim UI regardless of if there is a `hash` here?

Reply:
Oh yeah, good call no need to check for the hash